File: Data Structures & Algorithms/binary-search/submission-3.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1

        def condition(idx):
            return nums[idx] >= target

        while left < right:
            mid = left + (right - left) // 2
            
            if nums[mid] >= target:
                right = mid
            else:
                left = mid + 1

        return -1 if nums[left] != target else left

File: Data Structures & Algorithms/search-2d-matrix/submission-3.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rows = matrix.length, cols = matrix[0].length;

        int left = 0, right = rows * cols - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int valueAtMid = matrix[mid / cols][mid % cols];
            if (valueAtMid < target) {
                left = mid + 1;
            } else if (valueAtMid > target) {
                right = mid - 1;
            } else {
                return true;
            }
        }

        return false;
    }
}
